File: seed_core/seed_core/report/seed_stock_balance/seed_stock_balance.js
frappe.query_reports["Seed Stock Balance"] = {
    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company")
        },
        {
            "fieldname": "warehouse",
            "label": __("Warehouse"),
            "fieldtype": "Link",
            "options": "Warehouse"
        },
        {
            "fieldname": "crop",
            "label": __("Crop"),
            "fieldtype": "Link",
            "options": "Seed Crop"
        },
        {
            "fieldname": "segment",
            "label": __("Segment"),
            "fieldtype": "Link",
            "options": "Seed Segment",
            "get_query": function () {
                var crop = frappe.query_report.get_filter_value("crop");
                if (crop) {
                    return {
                        filters: { seed_crop: crop }
                    };
                }
            }
        },
        {
            "fieldname": "variety",
            "label": __("Variety"),
            "fieldtype": "Link",
            "options": "Seed Variety"
        },
        {
            "fieldname": "show_organic_only",
            "label": __("Organic Only"),
            "fieldtype": "Check"
        },
        {
            "fieldname": "show_untreated_only",
            "label": __("Untreated Only"),
            "fieldtype": "Check"
        },
        {
            "fieldname": "show_retest_due",
            "label": __("Retest Due Only"),
            "fieldtype": "Check"
        }
    ],
    "formatter": function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        // Highlight retest due batches
        if (column.fieldname === "status") {
            if (data.status === "Retest Due") {
                value = "<span style='color: red; font-weight: bold;'>" + value + "</span>";
            } else if (data.status === "Retest Soon") {
                value = "<span style='color: orange;'>" + value + "</span>";
            } else if (data.status === "OK") {
                value = "<span style='color: green;'>" + value + "</span>";
            }
        }

        return value;
    }
};

File: seed_core/seed_core/report/seed_stock_balance/seed_stock_balance.py
# Copyright (c) 2026, aremtech and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, today


def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	return columns, data


def get_columns():
	return [
		{
			"label": _("Crop"),
			"fieldname": "crop",
			"fieldtype": "Link",
			"options": "Seed Crop",
			"width": 120
		},
		{
			"label": _("Segment"),
			"fieldname": "segment",
			"fieldtype": "Link",
			"options": "Seed Segment",
			"width": 150
		},
		{
			"label": _("Variety"),
			"fieldname": "variety",
			"fieldtype": "Link",
			"options": "Seed Variety",
			"width": 150
		},
		{
			"label": _("Batch"),
			"fieldname": "batch",
			"fieldtype": "Link",
			"options": "Batch",
			"width": 120
		},
		{
			"label": _("Warehouse"),
			"fieldname": "warehouse",
			"fieldtype": "Link",
			"options": "Warehouse",
			"width": 150
		},
		{
			"label": _("Qty"),
			"fieldname": "qty",
			"fieldtype": "Float",
			"width": 100
		},
		{
			"label": _("Germination %"),
			"fieldname": "germination_percent",
			"fieldtype": "Percent",
			"width": 110
		},
		{
			"label": _("Purity %"),
			"fieldname": "purity_percent",
			"fieldtype": "Percent",
			"width": 90
		},
		{
			"label": _("Test Date"),
			"fieldname": "lab_test_date",
			"fieldtype": "Date",
			"width": 100
		},
		{
			"label": _("Retest Date"),
			"fieldname": "next_retest_date",
			"fieldtype": "Date",
			"width": 100
		},
		{
			"label": _("Organic"),
			"fieldname": "is_organic",
			"fieldtype": "Check",
			"width": 70
		},
		{
			"label": _("Treated"),
			"fieldname": "is_treated",
			"fieldtype": "Check",
			"width": 70
		},
		{
			"label": _("Status"),
			"fieldname": "status",
			"fieldtype": "Data",
			"width": 100
		}
	]


def get_data(filters):
	conditions = get_conditions(filters)

	# Get stock ledger entries grouped by batch and warehouse
	data = frappe.db.sql("""
		SELECT
			sv.seed_crop as crop,
			sv.seed_segment as segment,
			sv.name as variety,
			b.name as batch,
			sle.warehouse,
			SUM(sle.actual_qty) as qty,
			b.germination_percent,
			b.purity_percent,
			b.lab_test_date,
			b.next_retest_date,
			b.is_organic,
			CASE WHEN b.is_chemically_treated = 1 OR b.is_pelleted = 1 OR b.is_primed = 1 OR b.is_coated = 1 THEN 1 ELSE 0 END as is_treated
		FROM `tabStock Ledger Entry` sle
		JOIN `tabBatch` b ON sle.batch_no = b.name
		JOIN `tabItem` i ON sle.item_code = i.name
		LEFT JOIN `tabSeed Variety` sv ON sv.linked_item = i.name
		WHERE sle.is_cancelled = 0
		{conditions}
		GROUP BY sle.batch_no, sle.warehouse
		HAVING SUM(sle.actual_qty) > 0
		ORDER BY sv.seed_crop, sv.seed_segment, sv.name, b.name, sle.warehouse
	""".format(conditions=conditions), filters, as_dict=True)

	# Add status based on retest date
	today_date = getdate(today())
	for row in data:
		if row.next_retest_date:
			if getdate(row.next_retest_date) < today_date:
				row["status"] = "Retest Due"
			elif (getdate(row.next_retest_date) - today_date).days <= 30:
				row["status"] = "Retest Soon"
			else:
				row["status"] = "OK"
		else:
			row["status"] = "No Test"

	return data


def get_conditions(filters):
	conditions = ""

	if filters.get("company"):
		conditions += " AND sle.company = %(company)s"

	if filters.get("warehouse"):
		conditions += " AND sle.warehouse = %(warehouse)s"

	if filters.get("crop"):
		conditions += " AND sv.seed_crop = %(crop)s"

	if filters.get("segment"):
		conditions += " AND sv.seed_segment = %(segment)s"

	if filters.get("variety"):
		conditions += " AND sv.name = %(variety)s"

	if filters.get("show_organic_only"):
		conditions += " AND b.is_organic = 1"

	if filters.get("show_untreated_only"):
		conditions += " AND b.is_chemically_treated = 0 AND b.is_pelleted = 0 AND b.is_primed = 0 AND b.is_coated = 0"

	if filters.get("show_retest_due"):
		conditions += f" AND b.next_retest_date < '{today()}'"

	return conditions
